### PR TITLE
Center rolling window for time offset

### LINKED ISSUE #20012: Center rolling window with date NotImplementedError
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
import numpy as np

idx = pd.date_range(start="2010-01-01", end="2010-01-30", fred="1D")
df = pd.DataFrame(index=idx, data=np.sin(np.linspace(0,2*np.pi,len(idx))))

# works well
df.rolling(5, center=True, min_periods=5)

# fail with 
# NotImplementedError: center is not implemented for datetimelike and offset based windows
df.rolling("5D", center=True, min_periods=5)
```

Complete traceback:
<details>

```python

NotImplementedError                       Traceback (most recent call last)
<ipython-input-54-058f86b11afc> in <module>()
----> 1 df.rolling("5D", center=True, min_periods=5)

~/.local/lib/python3.5/site-packages/pandas/core/generic.py in rolling(self, window, min_periods, freq, center, win_type, on, axis, closed)
   7063                                    min_periods=min_periods, freq=freq,
   7064                                    center=center, win_type=win_type,
-> 7065                                    on=on, axis=axis, closed=closed)
   7066 
   7067         cls.rolling = rolling

~/.local/lib/python3.5/site-packages/pandas/core/window.py in rolling(obj, win_type, **kwds)
   2064         return Window(obj, win_type=win_type, **kwds)
   2065 
-> 2066     return Rolling(obj, **kwds)
   2067 
   2068 

~/.local/lib/python3.5/site-packages/pandas/core/window.py in __init__(self, obj, window, min_periods, freq, center, win_type, axis, on, closed, **kwargs)
     84         self.win_freq = None
     85         self.axis = obj._get_axis_number(axis) if axis is not None else None
---> 86         self.validate()
     87 
     88     @property

~/.local/lib/python3.5/site-packages/pandas/core/window.py in validate(self)
   1104             # we don't allow center
   1105             if self.center:
-> 1106                 raise NotImplementedError("center is not implemented "
   1107                                           "for datetimelike and offset "
   1108                                           "based windows")

NotImplementedError: center is not implemented for datetimelike and offset based windows

```

</details>

#### Problem description

The rolling window is not able to be centered with a datetimelike index, which is a problem when dealing with time series. 

I think that this is a WIP feature. I don't see anything blocking about it.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.0-3-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: C.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.21.0
pytest: 3.0.6
pip: 9.0.1
setuptools: 38.2.3
Cython: None
numpy: 1.13.3
scipy: 0.19.1
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: None
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: None
tables: 3.3.0
numexpr: 2.6.1
feather: None
matplotlib: 2.0.0
openpyxl: 2.3.0
xlrd: 1.0.0
xlwt: None
xlsxwriter: None
lxml: 3.7.1
bs4: 4.5.3
html5lib: 0.999999999
sqlalchemy: 1.2.2
pymysql: None
psycopg2: 2.6.2 (dt dec pq3 ext lo64)
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by luholgit**
- [x] closes #20012
- [x] tests added / passed
- [x] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Added center functionality for `VariableWindowIndexer`. Note - I am unsure if the NotImplementedError in lines 1966-1969 in rolling.py still correctly raises an error for offset based windows.

Finalizes previous PR https://github.com/pandas-dev/pandas/pull/36097

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
needs tests

**Comment by luholgit** [comment]
take

File: pandas/_libs/window/indexers.pyx [hunk 1]
```diff
@@ -11,7 +11,7 @@ def calculate_variable_window_bounds(
     int64_t num_values,
     int64_t window_size,
     object min_periods,  # unused but here to match get_window_bounds signature
-    object center,  # unused but here to match get_window_bounds signature
+    object center,
```
Comment (mroeschke) [inline pandas/_libs/window/indexers.pyx:14]: Should be `bint center` now if supported

File: pandas/core/window/rolling.py [hunk 2]
```diff
@@ -450,7 +457,36 @@ def homogeneous_func(values: np.ndarray):
             if values.size == 0:
                 return values.copy()
 
+            def _calculate_center_offset(window) -> int:
```
Comment (mroeschke) [inline pandas/core/window/rolling.py:460]: None of this code below should be needed.

The `center` calculation should just adjust the indexers return in `calculate_variable_window_bounds` and everything else should work correctly.

File: pandas/core/window/rolling.py [hunk 3]
```diff
@@ -334,6 +334,13 @@ def _get_window_indexer(self) -> BaseIndexer:
         """
         if isinstance(self.window, BaseIndexer):
             return self.window
+
+        if self.is_freq_type:
```
Comment (mroeschke) [inline pandas/core/window/rolling.py:338]: This clause shouldn't be needed either. `self,center` should just be passed to  `VariableWindowIndexer` in the clause below

File: pandas/core/window/rolling.py [hunk 4]
```diff
@@ -1880,18 +1925,18 @@ def validate(self):
...
             or isinstance(
                 self._on, (ABCDatetimeIndex, ABCTimedeltaIndex, ABCPeriodIndex)
             )
         ) and isinstance(self.window, (str, BaseOffset, timedelta)):
 
             self._validate_monotonic()
 
-            # we don't allow center
-            if self.center:
+            # we don't allow center for offset based windows
```
Comment (mroeschke) [inline pandas/core/window/rolling.py:1936]: This entire clause should be removed once `center` is supported

File: pandas/core/window/rolling.py [hunk 5]
```diff
@@ -450,7 +457,36 @@ def homogeneous_func(values: np.ndarray):
...
+                """
+                if not is_integer(window):
+                    window = len(window)
+                return int((window - 1) / 2.0)
+
+            offset = (
+                _calculate_center_offset(self.window)
+                if self.center
+                and not isinstance(
+                    self._get_window_indexer(self.window), VariableWindowIndexer
```
Comment (arw2019) [inline pandas/core/window/rolling.py:481]: I don't think you should be passing an argument here, just `self._get_window_indexer()`

File: pandas/_libs/window/indexers.pyx (hunk 1 above)
Comment (luholgit) [reply pandas/_libs/window/indexers.pyx]: thanks for your quick response, this is done now. :)

File: pandas/core/window/rolling.py (hunk 3 above)
Comment (luholgit) [reply pandas/core/window/rolling.py]: this is corrected too

File: pandas/core/window/rolling.py (hunk 5 above)
Comment (luholgit) [reply pandas/core/window/rolling.py]: I removed the argument here, too. Thanks for the feedback!

File: pandas/core/window/rolling.py (hunk 4 above)
Comment (luholgit) [reply pandas/core/window/rolling.py:1930]: removed this now, too! :)

File: pandas/_libs/window/indexers.pyx [hunk 6]
```diff
@@ -74,14 +78,27 @@ def calculate_variable_window_bounds(
     # right endpoint is open
     else:
         end[0] = 0
+    if center_window:
+        for j in range(0, num_values+1):
+            if (index[j] == index[0] + index_growth_sign*window_size/2 and
```
Comment (jreback) [inline pandas/_libs/window/indexers.pyx:83]: use spaces around operators & parens to format this (we don't lint these files automatically)

**Review by jreback — CHANGES_REQUESTED** [review]
you need tests to validate each path that is added.

File: pandas/core/window/rolling.py [hunk 7]
```diff
@@ -412,6 +415,19 @@ def _apply_tablewise(
...
+        """
+        if self.axis > result.ndim - 1:
+            raise ValueError("Requested axis is larger then no. of argument dimensions")
+
+        if offset > 0:
+            lead_indexer = [slice(None)] * result.ndim
+            lead_indexer[self.axis] = slice(offset, None)
+            result = np.copy(result[tuple(lead_indexer)])
+        return result
+
```
Comment (luholgit) [inline pandas/core/window/rolling.py:430]: In an effort to make the tests run again before introducing new ones, I encountered an error with "'Rolling' object has no attribute '_center_window'".
I tried to solve the error by adding the `_center_window()` method to the `BaseWindow` class. 
However, this resulted in new errors.

I am not sure whether it is correct to follow this debugging path. Could you quickly comment @jreback , @mroeschke  or @arw2019? 

Meanwhile, I am working on adding the changes you asked for. :)

Comment (mroeschke) [reply pandas/core/window/rolling.py:430]: As mentioned in https://github.com/pandas-dev/pandas/pull/38780#discussion_r549834716, I think you shouldn't need any additional code in this `_apply` method.

The changes should just need to occur in `pandas/_libs/window/indexers.pyx` in the `calculate_variable_window_bounds` function to make it return indexes that "center" the window.

Comment (mroeschke) [reply pandas/core/window/rolling.py:430]: For example, for non time offests, center is handled when calculating the indexes of each window

https://github.com/pandas-dev/pandas/blob/67d4cae17bec45e84b9cf51bcf4fb5bbe293b26f/pandas/core/window/indexers.py#L81

Comment (luholgit) [reply pandas/core/window/rolling.py]: > 
> 
> As mentioned in [#38780 (comment)](https://github.com/pandas-dev/pandas/pull/38780#discussion_r549834716), I think you shouldn't need any additional code in this `_apply` method.
> 
> The changes should just need to occur in `pandas/_libs/window/indexers.pyx` in the `calculate_variable_window_bounds` function to make it return indexes that "center" the window.

Just removed the unnecessary lines from rolling.py and that made the tests work again, thanks for that hint! At first I understood I had to migrate those lines to the `pandas/_libs/window/indexers.pyx` instead of ditching them all together.

File: pandas/_libs/window/indexers.pyx [hunk 8]
```diff
@@ -60,6 +62,8 @@ def calculate_variable_window_bounds(
 
     if index[num_values - 1] < index[0]:
         index_growth_sign = -1
+    if center:
```
Comment (mroeschke) [inline pandas/_libs/window/indexers.pyx:65]: Sine we typed `center` as `bint`. This isnt needed anymore

File: pandas/core/window/rolling.py [hunk 9]
```diff
@@ -334,9 +334,12 @@ def _get_window_indexer(self) -> BaseIndexer:
         """
         if isinstance(self.window, BaseIndexer):
             return self.window
+
```
Comment (mroeschke) [inline pandas/core/window/rolling.py:337]: Can you undo this white space?

File: pandas/core/window/rolling.py [hunk 10]
```diff
@@ -451,6 +454,7 @@ def homogeneous_func(values: np.ndarray):
                 return values.copy()
 
             def calc(x):
+
```
Comment (mroeschke) [inline pandas/core/window/rolling.py:457]: Can you undo this white space?

File: pandas/core/window/rolling.py [hunk 11]
```diff
@@ -1880,20 +1884,14 @@ def validate(self):
         # we allow rolling on a datetimelike index
         if (
             self.obj.empty
+            # TODO: add "or self.is_datetimelike"?
```
Comment (mroeschke) [inline pandas/core/window/rolling.py:1887]: Don't need this TODO

**Review by jreback — CHANGES_REQUESTED** [review]
@luholgit we need actual tests that lock down whether the bounds are correct including variable size windows. There *are* some center tests for fixed windows, follow those examples.

**Comment by luholgit** [comment]
Unfortunately I am currently sunk in work, but I will come back to implement the tests in the next week! Thanks for all the feedback so far. 👍

**Comment by luholgit** [comment]
@jreback we implemented testing of the new "center" to two already existing tests now. Is that what you had in mind?

File: pandas/_libs/window/indexers.pyx [hunk 12]
```diff
@@ -74,14 +76,27 @@ def calculate_variable_window_bounds(
     # right endpoint is open
     else:
         end[0] = 0
+    if center_window:
+        for j in range(0, num_values+1):
```
Comment (mroeschke) [inline pandas/_libs/window/indexers.pyx:80]: `range(num_values + 1)`

File: pandas/_libs/window/indexers.pyx [hunk 13]
```diff
@@ -74,14 +76,27 @@ def calculate_variable_window_bounds(
...
+            if (index[j] == index[0] + index_growth_sign * window_size / 2 and
```
Comment (mroeschke) [inline pandas/_libs/window/indexers.pyx:81]: Can we put all this `if` statement in 1 line?

File: pandas/_libs/window/indexers.pyx [hunk 14]
```diff
@@ -74,14 +76,27 @@ def calculate_variable_window_bounds(
     # right endpoint is open
     else:
         end[0] = 0
+    if center_window:
```
Comment (mroeschke) [inline pandas/_libs/window/indexers.pyx:79]: This should just be `center` right (and elsewhere)?

File: pandas/tests/window/test_rolling.py [hunk 15]
```diff
@@ -121,35 +121,58 @@ def test_numpy_compat(method):
         getattr(r, method)(dtype=np.float64)
 
 
-def test_closed_fixed(closed, arithmetic_win_operators):
+# maybe this one too?
+@pytest.mark.parametrize(
+    "center",
```
Comment (mroeschke) [inline pandas/tests/window/test_rolling.py:126]: I think we have a pytest fixture for this already.

File: pandas/tests/window/test_rolling.py [hunk 16]
```diff
@@ -121,35 +121,58 @@ def test_numpy_compat(method):
...
     )
-
-    rolling = df.rolling(window=len(df), closed="left", min_periods=1)
-    result = rolling.mean()
     expected = DataFrame(
-        [np.nan, 0, 0.5, 2 / 3, 0.5, 0.4, 0.5, 0.428571],
+        expected_data,
         columns=["binary_col"],
-        index=date_range(start="2020-01-01", freq="min", periods=len(data)),
+        index=date_range(start="2020-01-01", freq="min", periods=8),
```
Comment (mroeschke) [inline pandas/tests/window/test_rolling.py:171]: `periods=len(expected_data)`

File: pandas/tests/window/test_rolling.py [hunk 17]
```diff
@@ -394,6 +417,7 @@ def test_rolling_datetime(axis_frame, tz_naive_fixture):
     tm.assert_frame_equal(result, expected)
 
 
+# maybe this? center true/false
```
Comment (mroeschke) [inline pandas/tests/window/test_rolling.py:420]: Yes if you could test center here too that'd be great.

File: pandas/_libs/window/indexers.pyx (hunk 13 above)
Comment (luholgit) [reply pandas/_libs/window/indexers.pyx:81]: Tried this, but seemed to be an issue with the max character count per line (pre-commit), so changed this back again now

image

**Comment by sevberg** [comment]
Hey @mroeschke, 

We've encountered some unexpected behavior with the test "test_closed_fixed" (in pandas/tests/window/test_rolling.py), and before we make any substantial changes to the related functions (which are somewhat out of context of this PR, in our opinion), then we wanted to get your feedback.

As you can see in the screenshot below, when using the regular "rolling" functionality (i.e. without a time-index) and `center=True` we are not getting the result we would expect. It seems like the results are shifted by one index, and that there is a NaN value at the zeroth index. You can see this clearly when invoking a call to rolling with a `window=2`, `center=True`,  `closed="neither"`,  `min_periods=1`, and aggregating with `.mean()`. Then this should exactly reproduce the original Series. As you can see this works as expected when using a time-index (which we have just implemented), but does not appear to work otherwise. 

image

Maybe I am missing something simple here? Or is this indicative of a bug in the non-time-index rolling functionalities? If the latter, this is probably something that should be tackled in a separate PR?

File: pandas/tests/window/test_rolling.py (hunk 15 above)
Comment (luholgit) [reply pandas/tests/window/test_rolling.py]: Good hint, thanks! We used the fixture now instead.

File: pandas/tests/window/test_rolling.py (hunk 17 above)
Comment (luholgit) [reply pandas/tests/window/test_rolling.py:405]: added the `center=True` path in the same way now, and used the center fixture for it :)

**Comment by mroeschke** [comment]
@sevberg you bring up an interesting edge case. I think the fundamental question you're asking is "When centering a window with 2 points (or any even number of values), should the left edge or the right edge be considered the "center" "

I suppose there is a valid case for picking either, but my best answer is that it might be "pandas convention" that the left edge was chosen. I would recommend for the PR trying to match what the fixed window behavior does and if you disagree with the existing behavior you can open up a separate issue to discuss.

For reference, here's how the window edges for fixed window lengths are calculated: https://github.com/pandas-dev/pandas/blob/9c33e5e4f4c5dd6941da866234e506dae7d9f1e0/pandas/core/window/indexers.py#L81

**Comment by sevberg** [comment]
> For reference, here's how the window edges for fixed window lengths are calculated:
> 
> https://github.com/pandas-dev/pandas/blob/9c33e5e4f4c5dd6941da866234e506dae7d9f1e0/pandas/core/window/indexers.py#L81

Thanks a lot for pointing that out, @mroeschke. I completely understand your point about how this is an arbitrary decision when working in "index space", and why it's currently implemented in this way :).  As you suggested, I've tried to align the datetimelike-window behavior with the regular index-window but, after trying to think it through to ensure a general solution, I've ran into a few logical inconstancies which may make it impossible. In the end, I think it may be both easier and more "correct" to have the rolling functionality behave slightly differently in the "index-window" versus the "datetimelike-window" contexts when using a centered window. Hopefully I can convince you of this as well :)

The differences fundamentally stems from the consequence of working in "index space", where the limitation you described above applies, versus that of "continuous space" such as for a date-time-index or float-valued-index where it does not. In the latter case, it is always possible to define a window (lets call this "W") centered on some value ("X") by calculating X-W/2 to X+W/2, and this window will always be valid regardless if W is an even number (or even a whole number). If I wanted to make the centered-datetimelike-rolling behavior the same as the "index space" behavior, then I would also have to decide on what the smallest possible time-unit is ("U"). Afterwards I would be able to use this to determine if the given window is "effectively" even or odd, and then finally shift the window in the even case (via X-W/2+U/2 to X+W/2+U/2 ). But since we're working in "continuous space", then there is no such smallest time-unit. So considering that adding this shifting is not necessary (since the definition of the window is valid without it), would complicate the behavior of the function, and would be based on an arbitrary assumption, then I think it would be better not to include it at all.

Hopefully you see this the same way I do. If so, I've already added a new test to specifically test rolling functions for centered datetimelike windows.

Either way, I'm looking forward to your feedback!

**Comment by sevberg** [comment]
Also, I've identified a small bug (a real one, this time :p) in the way the `sem` function is calculated. In short, in the "pandas/core/nanops" functions you can see that the `ddof` variable is not subtracted from the array count, where as in the rolling version of the same function, it is included. I assume that the rolling version of this function should match the "global" one, yes? If so, this is already corrected :)

File: pandas/core/window/rolling.py [hunk 18]
```diff
@@ -1661,7 +1663,7 @@ def skew(self, **kwargs):
     )
 
     def sem(self, ddof: int = 1, *args, **kwargs):
-        return self.std(*args, **kwargs) / (self.count() - ddof).pow(0.5)
+        return self.std(*args, **kwargs) / self.count().pow(0.5)
```
Comment (jreback) [inline pandas/core/window/rolling.py:1666]: don't you need to pass ddof?

**Review by jreback — CHANGES_REQUESTED** [review]
pls add a whatsnew note in Enh for 1.3. Also see if any docs need updating in window.rst

**Comment by mroeschke** [comment]
@sevberg I agree with your explanation in theory but

> If I wanted to make the centered-datetimelike-rolling behavior the same as the "index space" behavior, then I would also have to decide on what the smallest possible time-unit is ("U")

"U" in this case is 1 nanosecond as the cython code that determines the window bounds for rolling datetimes deals with int64 representations of the times (epoch nanoseconds).

File: pandas/core/window/rolling.py (hunk 18 above)
Comment (sevberg) [reply pandas/core/window/rolling.py:1666]: I would say yes, ddof should be passed. But when comparing the rolling (and also expanding) version of the sem function with that of the 'regular' sem (i.e. nansem) function which is invoked (i.e. when called from a dataframe or series), you can see that the behavior is different:
image

Digging in to it, you can see that in the [nansem version of the function](https://github.com/pandas-dev/pandas/blob/8e2e98baa62de4629ed3240abb66be9cc1c3ead5/pandas/core/nanops.py#L941) does not pass ddof, while [the rolling version](https://github.com/pandas-dev/pandas/blob/8e2e98baa62de4629ed3240abb66be9cc1c3ead5/pandas/core/window/rolling.py#L1664) does. In a perfect world I would say that passing ddof is correct, and that the nansem function has a problem. But, at the same time, I think the nansem version of the one which is used more regularly and is more heavily "tested", and thus the behavior of the rolling sem should at least match it.

Please let me know if you see it differently :)

**Comment by sevberg** [comment]
hello again @mroeschke, you're totally right to point that out. But I guess it doesn't make much of a difference for the [test in question](https://github.com/pandas-dev/pandas/blob/8e2e98baa62de4629ed3240abb66be9cc1c3ead5/pandas/tests/window/test_rolling.py#L124-L135) where the time-step is in days. So do I understand correctly that it is alright to move forward with testing centered datetimelike-windows separately from index-widows?

File: pandas/core/window/rolling.py (hunk 18 above)
Comment (sevberg) [reply pandas/core/window/rolling.py]: update: Even though this is an inconsistency between the two versions of the 'sem' function, makeing this ddof change in either the rolling version, or the nanops version leads to a rabbit-hole of related tests that need fixing, all of which are out-of-context of datetimelike-window centering 🤦. Because of that, this is probably something which should be tackled in a follow-up issue. Therefore I've reverted to the previous state, and added a work-around to the test. Hope this makes sense!

**Comment by sevberg** [comment]
> pls add a whatsnew note in Enh for 1.3. Also see if any docs need updating in window.rst

Hello @jreback, I've added a, enhancement note in whatsnew (v1.3), but I did not see something in window.rst which obviously needed attention

**Comment by mroeschke** [comment]
@sevberg true given that test in question, having a different result for time based rolling windows is okay.

Could you also add test case though where that +/- 1 nanosecond does matter?

File: pandas/tests/window/test_rolling.py [hunk 19]
```diff
@@ -130,29 +130,94 @@ def test_closed_fixed(closed, arithmetic_win_operators):
     df_fixed = DataFrame({"A": [0, 1, 2, 3, 4]})
     df_time = DataFrame({"A": [0, 1, 2, 3, 4]}, index=date_range("2020", periods=5))
 
-    result = getattr(df_fixed.rolling(2, closed=closed, min_periods=1), func_name)()
-    expected = getattr(df_time.rolling("2D", closed=closed), func_name)().reset_index(
-        drop=True
-    )
+    result = getattr(
```
Comment (jreback) [inline pandas/tests/window/test_rolling.py:133]: why is this needed? center is default False

File: pandas/tests/window/test_rolling.py [hunk 20]
```diff
@@ -130,29 +130,94 @@ def test_closed_fixed(closed, arithmetic_win_operators):
...
+    )()
+    expected = getattr(
+        df_time.rolling("2D", closed=closed, min_periods=1, center=False), func_name
+    )().reset_index(drop=True)
 
     tm.assert_frame_equal(result, expected)
 
 
-def test_closed_fixed_binary_col():
+def test_datetimelike_centered_selections(closed, arithmetic_win_operators):
```
Comment (jreback) [inline pandas/tests/window/test_rolling.py:143]: can you paramterize on closed (e.g. directly put your window selection up there with the closed parameter),  eg.. don't use the fixture

File: pandas/tests/window/test_rolling.py [hunk 21]
```diff
@@ -408,53 +473,97 @@ def test_rolling_window_as_string(using_array_manager):
...
+            + [98.0] * 21
+            + [95.0] * 14
+            + [91.0] * 7
+            + [86.0]
+            + [93.0] * 3
+            + [95.0] * 29
+            + [77.0] * 2
+        )
+
+    else:
```
Comment (jreback) [inline pandas/tests/window/test_rolling.py:521]: if doing this pls paramterize outside the function

**Review by jreback — CHANGES_REQUESTED** [review]
pls also add a whatsnew note (enh is ok), as well as see if any updates are needed in the https://pandas.pydata.org/pandas-docs/dev/user_guide/window.html, also the doc-string of rolling might need a small update as well

File: doc/source/whatsnew/v1.3.0.rst [hunk 22]
```diff
@@ -134,6 +134,30 @@ a copy will no longer be made (:issue:`32960`)
...
 
+Centered Datetime-Like Rolling Windows
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+When performing rolling calculations on :class:`DataFrame` and :class:`Series`
+objects with a datetime-like index, a centered datetime-like window can now be
+used (:issue:`38780`).
+For example:
+
+.. code-block:: ipython
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:145]: use `ipython:: python` so the code runs (don't need the prompts or output just the code)

File: doc/source/whatsnew/v1.3.0.rst [hunk 23]
```diff
@@ -134,6 +134,30 @@ a copy will no longer be made (:issue:`32960`)
 The default behavior when not passing ``copy`` will remain unchanged, i.e.
 a copy will be made.
 
+Centered Datetime-Like Rolling Windows
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:137]: would not object to an example in window.rst as well

File: pandas/_libs/window/indexers.pyx [hunk 24]
```diff
@@ -33,7 +33,7 @@ def calculate_variable_window_bounds(
         ignored, exists for compatibility
 
     center : object
-        ignored, exists for compatibility
+        center the rolling window on the current observation
```
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/_libs/window/indexers.pyx:36]: can you update

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/window/indexers.pyx (hunk 24 above)
[the remark this answers is not in the document]
Comment (luholgit) [reply pandas/_libs/window/indexers.pyx:36]: Yes, you are right, I changed this!

File: pandas/tests/window/test_rolling.py (hunk 20 above)
Comment (luholgit) [reply pandas/tests/window/test_rolling.py]: Good idea, this is done, too.

File: pandas/tests/window/test_rolling.py (hunk 21 above)
Comment (luholgit) [reply pandas/tests/window/test_rolling.py]: Done!

File: doc/source/whatsnew/v1.3.0.rst [hunk 25]
```diff
@@ -134,6 +134,30 @@ a copy will no longer be made (:issue:`32960`)
...
+For example:
+
+.. ipython:: python
+
+   In [1]: df = DataFrame(
+              {"A": [0, 1, 2, 3, 4]},
+              index=date_range("2020", periods=5, freq="1D")
+           )
+
+   In [2]: df.rolling("2D", center=True).mean()
```
Comment (mroeschke) [inline doc/source/whatsnew/v1.3.0.rst:152]: Don't need the `In` `Out` annotations when using `.. ipython:: python` (the code will execute line by line

File: pandas/tests/window/test_rolling.py (hunk 19 above)
Comment (luholgit) [reply pandas/tests/window/test_rolling.py:135]: I removed this redundand "False". :)

File: doc/source/whatsnew/v1.3.0.rst (hunk 23 above)
Comment (luholgit) [reply doc/source/whatsnew/v1.3.0.rst:137]: I added the same example also to the window.rst now

File: pandas/tests/window/test_rolling.py [hunk 26]
```diff
@@ -130,29 +132,125 @@ def test_closed_fixed(closed, arithmetic_win_operators):
...
+    s = Series(range(3), index=date_range(start="2020-01-01", freq="D", periods=3))
+
+    # behavior of index- and datetime-based windows differs here!
+    # s.rolling(window=2, min_periods=1, center=True).mean()
+
+    result = s.rolling(window="2D", min_periods=1, center=True).mean()
+
+    expected = Series([0.5, 1.5, 2], index=s.index)
+
+    tm.assert_series_equal(result, expected)
```
Comment (luholgit) [inline pandas/tests/window/test_rolling.py:230]: @mroeschke I added this test to check for the behavior when operating with even sized windows. As mentioned by @sevberg this currently does behave differently than the classic index-based windows.

Is that what you had in mind [here](https://github.com/pandas-dev/pandas/pull/38780#issuecomment-772766767)?

Comment (luholgit) [reply pandas/tests/window/test_rolling.py:230]: This test would also work with `N` instead of `D` as frequency, therefore I am not sure if this is what you meant when referring to nano seconds?

Comment (mroeschke) [reply pandas/tests/window/test_rolling.py:230]: Yes I believe this test exercises the difference in behavior which is good.

File: doc/source/user_guide/window.rst [hunk 27]
```diff
@@ -156,6 +156,16 @@ By default the labels are set to the right edge of the window, but a
...
 
+This can also be applied to datetime-like indices.
+
+.. ipython:: python
+
+    df = pd.DataFrame(
+        {"A": [0, 1, 2, 3, 4]}, index=pd.date_range("2020", periods=5, freq="1D")
+    )
+
+    df.rolling("2D", center=True).mean()
```
Comment (jreback) [inline doc/source/user_guide/window.rst:167]: can you make 2 ipython blocks, first should show df, second the df.rolling (i mean not a big deal)

File: doc/source/whatsnew/v1.3.0.rst [hunk 28]
```diff
@@ -134,6 +134,23 @@ a copy will no longer be made (:issue:`32960`)
...
+When performing rolling calculations on :class:`DataFrame` and :class:`Series`
+objects with a datetime-like index, a centered datetime-like window can now be
+used (:issue:`38780`).
+For example:
+
+.. ipython:: python
+
+    df = pd.DataFrame(
+        {"A": [0, 1, 2, 3, 4]}, index=pd.date_range("2020", periods=5, freq="1D")
+    )
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:149]: same comment as above

File: doc/source/user_guide/window.rst [hunk 29]
```diff
@@ -156,6 +156,16 @@ By default the labels are set to the right edge of the window, but a
    s.rolling(window=5).mean()
    s.rolling(window=5, center=True).mean()
 
+This can also be applied to datetime-like indices.
+
```
Comment (jreback) [inline doc/source/user_guide/window.rst:160]: can you add a versionadded 1.3 tag here

**Review by jreback — COMMENTED** [review]
just a minor doc comment, ping on green (I think you addressed all of @mroeschke comments but if not pls do)

File: pandas/tests/window/test_rolling.py (hunk 26 above)
Comment (luholgit) [reply pandas/tests/window/test_rolling.py:230]: Great, thanks!

File: doc/source/whatsnew/v1.3.0.rst (hunk 28 above)
Comment (luholgit) [reply doc/source/whatsnew/v1.3.0.rst:149]: Done.

File: doc/source/user_guide/window.rst (hunk 27 above)
Comment (luholgit) [reply doc/source/user_guide/window.rst:169]: I split this one and added the `.rolling(..., center=False)` line for clarity, as in the example above.

File: doc/source/user_guide/window.rst (hunk 29 above)
Comment (luholgit) [reply doc/source/user_guide/window.rst:159]: Done, too!

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @luholgit very nice!

**Comment by luholgit** [comment]
Big thanks to you, @jreback and @mroeschke! I've learned a lot while working on this. :)
